### PR TITLE
[CARBONDATA-3454] optimized index server output for count(*)

### PR DESCRIPTION
**Pull request by kunal642**
Optimised the output for count(*) queries so that only a long is send back to the driver to reduce the network transfer cost for index server

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11981/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3918/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3706/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11982/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3919/

File: core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlocklet.java [hunk 1]
```diff
@@ -161,30 +176,35 @@ public void setColumnSchema(List<ColumnSchema> columnSchema) {
    * @param uniqueLocation
    * @throws IOException
    */
-  public void serializeData(DataOutput out, Map<String, Short> uniqueLocation)
+  public void serializeData(DataOutput out, Map<String, Short> uniqueLocation, boolean isCountJob)
       throws IOException {
     super.write(out);
-    if (dataMapUniqueId == null) {
-      out.writeBoolean(false);
+    if (isCountJob) {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlocklet.java]: change this code to if(isCountJob) {//your code, return; }

Comment (kunal642) [reply core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlocklet.java]: this is a if else block, return is not required. i have added return to the read method instead.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3709/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11985/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3713/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3922/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11989/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3926/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3755/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3967/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/12031/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3766/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3978/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/12042/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3831/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/4036/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/12104/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3832/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/12105/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/4037/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3849/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/4054/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/12122/

**Comment by ravipesala** [comment]
retest this please

**Comment by ravipesala** [comment]
add to whitelist

**Comment by ravipesala** [comment]
add to whitelist

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/120/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/115/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/121/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/12138/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/4070/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/200/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/194/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/198/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/229/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/226/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/224/

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/235/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/233/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/230/

File: core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlockletWrapper.java [hunk 2]
```diff
@@ -248,4 +247,8 @@ private void writeBlockletToStream(DataOutputStream stream, byte[] data,
     }
     this.dataSize = in.readInt();
   }
+
+  public long getCount() {
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlockletWrapper.java]: Please remove if not used

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonInputFormat.java [hunk 3]
```diff
@@ -412,10 +413,41 @@ public static ReadCommittedScope getReadCommittedScope(Configuration configurati
    */
   @Override public abstract List<InputSplit> getSplits(JobContext job) throws IOException;
 
+  Long getDistributedCount(CarbonTable table,
```
Comment (ravipesala) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonInputFormat.java]: Add comments

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonInputFormat.java [hunk 4]
```diff
@@ -412,10 +413,41 @@ public static ReadCommittedScope getReadCommittedScope(Configuration configurati
...
+
   List<ExtendedBlocklet> getDistributedSplit(CarbonTable table,
       FilterResolverIntf filterResolverIntf, List<PartitionSpec> partitionNames,
       List<Segment> validSegments, List<Segment> invalidSegments,
       List<String> segmentsToBeRefreshed) throws IOException {
+    return getDistributedSplit(table, filterResolverIntf, partitionNames, validSegments,
+        invalidSegments, segmentsToBeRefreshed, false);
+  }
+
+  List<ExtendedBlocklet> getDistributedSplit(CarbonTable table,
```
Comment (ravipesala) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonInputFormat.java]: Change the name to reflect count for each block

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java [hunk 5]
```diff
@@ -562,15 +563,14 @@ public BlockMappingVO getBlockRowCount(Job job, CarbonTable table,
       if (CarbonProperties.getInstance()
           .isDistributedPruningEnabled(table.getDatabaseName(), table.getTableName())) {
         try {
-          List<InputSplit> extendedBlocklets = CarbonTableInputFormat.convertToCarbonInputSplit(
+          List<ExtendedBlocklet> extendedBlocklets =
               getDistributedSplit(table, null, partitions, filteredSegment,
-                  allSegments.getInvalidSegments(), toBeCleanedSegments));
-          for (InputSplit extendedBlocklet : extendedBlocklets) {
-            CarbonInputSplit blocklet = (CarbonInputSplit) extendedBlocklet;
+                  allSegments.getInvalidSegments(), toBeCleanedSegments);
```
Comment (ravipesala) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java]: I think you supposed to pass true for count job

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedCountRDD.scala [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+import org.apache.carbondata.common.logging.LogServiceFactory
+import org.apache.carbondata.core.cache.CacheProvider
+import org.apache.carbondata.core.datamap.{DataMapStoreManager, DistributableDataMapFormat}
+import org.apache.carbondata.core.datamap.dev.expr.DataMapDistributableWrapper
+import org.apache.carbondata.core.datastore.impl.FileFactory
+import org.apache.carbondata.core.util.{CarbonProperties, CarbonThreadFactory}
+import org.apache.carbondata.spark.rdd.CarbonRDD
+
+class DistributedCountRDD(@transient ss: SparkSession, dataMapFormat: DistributableDataMapFormat)
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedCountRDD.scala]: Add comment

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedCountRDD.scala [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+  override protected def getPreferredLocations(split: Partition): Seq[String] = {
+    if (split.asInstanceOf[DataMapRDDPartition].getLocations != null) {
+      split.asInstanceOf[DataMapRDDPartition].getLocations.toSeq
+    } else {
+      Seq()
+    }
+  }
+
+  private def groupSplits(xs: Seq[InputSplit], n: Int) = {
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedCountRDD.scala]: Extract to common method and use from prune RDD as well

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedCountRDD.scala [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    val (quot, rem) = (xs.size / n, xs.size % n)
+    val (smaller, bigger) = xs.splitAt(xs.size - rem * (quot + 1))
+    (smaller.grouped(quot) ++ bigger.grouped(quot + 1)).toList
+  }
+
+  override def internalCompute(split: Partition,
+      context: TaskContext): Iterator[(String, String)] = {
+    val attemptId = new TaskAttemptID(jobTrackerId, id, TaskType.MAP, split.index, 0)
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedCountRDD.scala]: Try to use the same code which was already present in prune RDD

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/257/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/253/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/254/

File: core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlockletWrapper.java (hunk 2 above)
Comment (kunal642) [reply core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlockletWrapper.java]: removed

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonInputFormat.java (hunk 4 above)
Comment (kunal642) [reply hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonInputFormat.java]: renamed

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java (hunk 5 above)
Comment (kunal642) [reply hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java]: done

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedCountRDD.scala (hunk 6 above)
Comment (kunal642) [reply integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedCountRDD.scala]: added

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedCountRDD.scala (hunk 8 above)
Comment (kunal642) [reply integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedCountRDD.scala]: extracted some parts but internalCompute logic is different

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/275/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/272/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/271/

**Comment by brijoobopanna** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/321/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/317/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/318/

File: core/src/main/java/org/apache/carbondata/core/datamap/AbstractDataMapJob.java [hunk 9]
```diff
@@ -30,4 +30,8 @@
   @Override public void execute(CarbonTable carbonTable,
       FileInputFormat<Void, BlockletDataMapIndexWrapper> format) {
   }
+
+  @Override public Long executeCountJob(DistributableDataMapFormat dataMapFormat) {
+    return null;
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/datamap/AbstractDataMapJob.java]: retrun 0 instead of null

Comment (kunal642) [reply core/src/main/java/org/apache/carbondata/core/datamap/AbstractDataMapJob.java]: removed the implementation in AbstractDataMapJob.java, now all the extending classes should give their own implementations

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/492/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/493/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/493/

**Comment by kumarvishal09** [comment]
LGTM, @kunal642 Please update the PR description

**Comment by kunal642** [comment]
Done

**Pull request closed without merging**